expression/builtin_string_vec.go

```diff
@@ -333,11 +333,64 @@ func (b *builtinConcatSig) vecEvalString(input *chunk.Chunk, result *chunk.Colum
 }
 
 func (b *builtinLocate3ArgsSig) vectorized() bool {
-	return false
+	return true
 }
 
+// vecEvalInt evals LOCATE(substr,str,pos), non case-sensitive.
+// See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_locate
 func (b *builtinLocate3ArgsSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalString(b.ctx, input, buf); err != nil {
+		return err
+	}
+	buf1, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalString(b.ctx, input, buf1); err != nil {
+		return err
+	}
+	// store positions in result
+	if err := b.args[2].VecEvalInt(b.ctx, input, result); err != nil {
+		return err
+	}
+
+	result.MergeNulls(buf, buf1)
+	i64s := result.Int64s()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
```

Comment:
Consider this case, b.args[2] is null but the result is not null.
```mysql
mysql> SELECT LOCATE('bar', 'foobarbar', NULL);
+----------------------------------+
| LOCATE('bar', 'foobarbar', NULL) |
+----------------------------------+
|                                0 |
+----------------------------------+
1 row in set (0.00 sec)
```

Reply:
My test result is correct

```sql
mysql> SELECT LOCATE('bar', 'foobarbar', NULL);
+----------------------------------+
| LOCATE('bar', 'foobarbar', NULL) |
+----------------------------------+
|                             NULL |
+----------------------------------+
1 row in set (0.00 sec)
```

Reply:
can args[0] and args[1] be null?

Reply:
yes, any args is null will result in null

Reply:
> Consider this case, b.args[2] is null but the result is not null.
> 
> ```sql
> mysql> SELECT LOCATE('bar', 'foobarbar', NULL);
> +----------------------------------+
> | LOCATE('bar', 'foobarbar', NULL) |
> +----------------------------------+
> |                                0 |
> +----------------------------------+
> 1 row in set (0.00 sec)
> ```

In MySQL 8.x, this SQL returns `NULL`.